library/spdm_responder_lib/libspdm_rsp_chunk_send.c

```diff
@@ -1,15 +1,244 @@
+                                                 size_t *response_size,
+                                                 void *response)
 {
-    LIBSPDM_ASSERT(false);
-    return LIBSPDM_STATUS_UNSUPPORTED_CAP;
+    const spdm_chunk_send_request_t *spdm_request;
+    spdm_chunk_send_ack_response_t *spdm_response;
+    libspdm_context_t *spdm_context;
+
+    libspdm_chunk_info_t *send_info;
+    libspdm_return_t status = LIBSPDM_STATUS_SUCCESS;
+
+    const uint8_t *chunk;
+    uint32_t large_message_size;
+    uint32_t calc_chunk_size;
+    uint8_t *scratch_buffer;
+    size_t scratch_buffer_size;
+    uint8_t* chunk_response;
+    size_t chunk_response_size;
+
+    spdm_context = context;
+    spdm_request = (const spdm_chunk_send_request_t*) request;
+
+    if ((spdm_context->local_context.capability.flags &
+         SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_CHUNK_CAP) == 0) {
+        return libspdm_generate_error_response(
+            spdm_context,
+            SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
+            response_size, response);
+    }
+
+    if (spdm_context->response_state != LIBSPDM_RESPONSE_STATE_NORMAL) {
+        return libspdm_responder_handle_response_state(
+            spdm_context,
+            spdm_request->header.request_response_code,
+            response_size, response);
+    }
+
+    if (spdm_context->connection_info.connection_state <
+        LIBSPDM_CONNECTION_STATE_AFTER_CAPABILITIES) {
+        libspdm_generate_error_response(
+            spdm_context, SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
+            response_size, response);
+        return LIBSPDM_STATUS_SUCCESS;
+    }
+
+    if (request_size < sizeof(spdm_chunk_send_request_t)) {
+        libspdm_generate_error_response(
+            spdm_context, SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+            response_size, response);
+        return LIBSPDM_STATUS_SUCCESS;
+    }
+
+    if (spdm_request->header.spdm_version < SPDM_MESSAGE_VERSION_12) {
+        return libspdm_generate_error_response(
+            spdm_context, SPDM_ERROR_CODE_UNSUPPORTED_REQUEST, 0,
+            response_size, response);
+    }
+
+    if (spdm_request->header.spdm_version
+        != libspdm_get_connection_version(spdm_context)) {
+        return libspdm_generate_error_response(
+            spdm_context, SPDM_ERROR_CODE_VERSION_MISMATCH, 0,
+            response_size, response);
+    }
+
+    if (spdm_context->chunk_context.get.chunk_in_use) {
+        /* Spec does not support simultanious chunk send and chunk get */
+        return libspdm_generate_error_response(
+            spdm_context, SPDM_ERROR_CODE_UNSUPPORTED_REQUEST, 0,
+            response_size, response);
+    }
+
+    send_info = &spdm_context->chunk_context.send;
+
+    if (!send_info->chunk_in_use) {
+
+        large_message_size = *(const uint32_t*) (spdm_request + 1);
+        chunk = (((const uint8_t*) (spdm_request + 1)) + sizeof(uint32_t));
+        calc_chunk_size =
```

Comment:
I think @jyao1 said that `request_size` is only an approximation due to transport data paths that don't have byte granularity. Ie, the data path is 8-bytes.

Reply:
Yes. Fixed.